File: scripts/test_solutions/295_find_median_data_stream.js
var MedianFinder = function() { this.lo = []; this.hi = []; };
MedianFinder.prototype.addNum = function(num) {
    this.lo.push(num); this.lo.sort((a, b) => b - a);
    this.hi.push(this.lo.shift()); this.hi.sort((a, b) => a - b);
    if (this.hi.length > this.lo.length) {
        this.lo.push(this.hi.shift());
        this.lo.sort((a, b) => b - a);
    }
};
MedianFinder.prototype.findMedian = function() {
    return this.lo.length > this.hi.length ? this.lo[0] : (this.lo[0] + this.hi[0]) / 2;
};

File: scripts/test_solutions/295_find_median_data_stream.cpp
class MedianFinder {
public:
    MedianFinder() {}

    void addNum(int num) {
        lo.push(num);
        hi.push(-lo.top());
        lo.pop();
        if (hi.size() > lo.size()) {
            lo.push(-hi.top());
            hi.pop();
        }
    }

    double findMedian() {
        if (lo.size() > hi.size()) {
            return lo.top();
        }
        return (lo.top() - hi.top()) / 2.0;
    }

private:
    priority_queue<int> lo;  // max heap
    priority_queue<int> hi;  // min heap (negated)
};
